File: app/src/main/java/antoninovitale/dropcodechallenge/details/section/model/mapper/IngredientSectionModelMapper.java
package antoninovitale.dropcodechallenge.details.section.model.mapper;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import antoninovitale.dropcodechallenge.api.model.Amount;
import antoninovitale.dropcodechallenge.api.model.Hop;
import antoninovitale.dropcodechallenge.api.model.Malt;
import antoninovitale.dropcodechallenge.details.section.model.Add;
import antoninovitale.dropcodechallenge.details.section.model.IngredientSectionModel;
import antoninovitale.dropcodechallenge.details.section.model.IngredientType;
import antoninovitale.dropcodechallenge.details.section.model.Status;

/**
 * Created by antoninovitale on 29/08/2017.
 */
public class IngredientSectionModelMapper {

    private static Comparator<IngredientSectionModel> ingredientSectionModelComparator = new
            Comparator<IngredientSectionModel>() {

                @Override
                public int compare(IngredientSectionModel model1, IngredientSectionModel model2) {
                    if (model1.getAdd() == model2.getAdd()) {
                        return 0;
                    }
                    switch (model1.getAdd()) {
                        case START:
                            return -1;
                        case MIDDLE: {
                            if (model2.getAdd() == Add.START) {
                                return 1;
                            } else {
                                return -1;
                            }
                        }
                        case END:
                            return 1;
                    }
                    return 0;
                }

            };

    private static IngredientSectionModel convertHop(Hop hop) {
        IngredientSectionModel model = null;
        if (hop != null) {
            model = new IngredientSectionModel(hop.getName(), convertAmount(hop.getAmount()),
                    IngredientType.HOP, Status.IDLE, Add.getAdd(hop.getAdd()));
        }
        return model;
    }

    private static IngredientSectionModel convertMalt(Malt malt) {
        IngredientSectionModel model = null;
        if (malt != null) {
            model = new IngredientSectionModel(malt.getName(), convertAmount(malt.getAmount()),
                    IngredientType.MALT, Status.IDLE, Add.NONE);
        }
        return model;
    }

    private static String convertAmount(Amount amount) {
        return amount != null ? String.format("%s %s", amount.getValue(), amount.getUnit()) : null;
    }

    public static List<IngredientSectionModel> convertHops(List<Hop> hops) {
        List<IngredientSectionModel> models = Collections.emptyList();
        if (hops != null && !hops.isEmpty()) {
            models = new ArrayList<>(hops.size());
            for (Hop hop : hops) {
                IngredientSectionModel model = convertHop(hop);
                if (model != null) {
                    models.add(model);
                }
            }
            if (!models.isEmpty()) {
                Collections.sort(models, ingredientSectionModelComparator);
            }
        }
        return models;
    }

    public static List<IngredientSectionModel> convertMalts(List<Malt> malts) {
        List<IngredientSectionModel> models = Collections.emptyList();
        if (malts != null && !malts.isEmpty()) {
            models = new ArrayList<>(malts.size());
            for (Malt malt : malts) {
                IngredientSectionModel model = convertMalt(malt);
                if (model != null) {
                    models.add(model);
                }
            }
            if (!models.isEmpty()) {
                Collections.sort(models, ingredientSectionModelComparator);
            }
        }
        return models;
    }

}

File: app/src/main/java/antoninovitale/punkapi/app/details/section/model/mapper/IngredientSectionModelMapper.kt
package antoninovitale.punkapi.app.details.section.model.mapper

import antoninovitale.punkapi.app.api.model.Amount
import antoninovitale.punkapi.app.api.model.Hop
import antoninovitale.punkapi.app.api.model.Malt
import antoninovitale.punkapi.app.details.section.model.Add
import antoninovitale.punkapi.app.details.section.model.Add.END
import antoninovitale.punkapi.app.details.section.model.Add.MIDDLE
import antoninovitale.punkapi.app.details.section.model.Add.NONE
import antoninovitale.punkapi.app.details.section.model.Add.START
import antoninovitale.punkapi.app.details.section.model.IngredientSectionModel
import antoninovitale.punkapi.app.details.section.model.IngredientType.HOP
import antoninovitale.punkapi.app.details.section.model.IngredientType.MALT
import antoninovitale.punkapi.app.details.section.model.Status.IDLE
import java.util.Comparator

/**
 * Created by antoninovitale on 29/08/2017.
 */
object IngredientSectionModelMapper {
    private val ingredientSectionModelComparator = Comparator { model1: IngredientSectionModel, model2: IngredientSectionModel ->
        if (model1.add == model2.add) return@Comparator 0
        return@Comparator when (model1.add) {
            START -> -1
            MIDDLE -> if (model2.add == START) 1 else -1
            END -> 1
            NONE -> 0
        }
    }

    private fun convertHop(hop: Hop?): IngredientSectionModel? {
        var model: IngredientSectionModel? = null
        if (hop != null) {
            model = IngredientSectionModel(hop.name, convertAmount(hop.amount), HOP, IDLE,
                    Add.getAdd(hop.add))
        }
        return model
    }

    private fun convertMalt(malt: Malt?): IngredientSectionModel? {
        var model: IngredientSectionModel? = null
        if (malt != null) {
            model = IngredientSectionModel(malt.name, convertAmount(malt.amount), MALT, IDLE, NONE)
        }
        return model
    }

    private fun convertAmount(amount: Amount?): String? {
        return if (amount != null) String.format("%s %s", amount.value, amount.unit) else null
    }

    fun convertHops(hops: List<Hop?>?): List<IngredientSectionModel> {
        if (hops != null && hops.isNotEmpty()) {
            return hops
                    .mapNotNull { hop -> convertHop(hop) }
                    .sortedWith(ingredientSectionModelComparator)
        }
        return emptyList()
    }

    fun convertMalts(malts: List<Malt?>?): List<IngredientSectionModel> {
        if (malts != null && malts.isNotEmpty()) {
            return malts
                    .mapNotNull { malt -> convertMalt(malt) }
                    .sortedWith(ingredientSectionModelComparator)
        }
        return emptyList()
    }
}